### PR TITLE
Supersimple/package search

### LINKED ISSUE #1256: Persist searches that return 0 results
When a package search term returns 0 results, store that string in a database.
The database should include the term searched for, the number of occurrences (frequency), and a reference to a package id.
usage:
1. if a term comes up frequently, eg. "lightstep", it is an indicator of a library the community is looking for in our ecosystem and could be used to solicit the community to build a library.
2. a frequent misspelling could redirect to a package automagically (eg. 'etco' could just redirect to 'ecto')

### PR DESCRIPTION
**Pull request by supersimple**
Closes #1256 

This PR adds persistence to failed searches; package searches that come back with 0 results.
Each time this occurs, we store the search term used in the database. If the term is searched for again (before the record is purged) we increment the frequency counter.
It also contains an endpoint to download a CSV of the 100 most frequent failed searches (requires a login)
There also includes a ~stored procedure~ release task to purge the infrequent searches after a period of time (set to 1 month). I expect the database to mostly fill up with one-off fails, which adds noise and little useful insights.

WHY?
Ultimately, I would like to find frequent misspellings of package names and use this data to redirect or suggest.
Also, the data could be used to reveal holes in the ecosystem.

NOTES:
The ~stored procedure~ release task will either need to be called manually, or more likely, be called using a cron. GCP has native tooling for this.

### CODE REVIEW

File: lib/hexpm/repository/package_searches/package_searches.ex [hunk 1]
```diff
@@ -0,0 +1,41 @@
+defmodule Hexpm.Repository.PackageSearches do
+  use Hexpm.Context
+  alias Hexpm.Repository.PackageSearches.PackageSearch
+
+  def add_or_increment(params) do
+    case get(params["term"]) do
+      nil -> add(%PackageSearch{}, params)
+      %PackageSearch{} = package_search -> increment(package_search)
+    end
+  end
```
Comment (ericmj) [inline lib/hexpm/repository/package_searches/package_searches.ex:10]: Let's do this as an upsert instead?

Something like this I think:

```elixir
Repo.insert(changeset, on_conflict: update(PackageSearch, [inc: [frequency: 1]])
```

File: test_suite_analysis/metadata.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"current_schema_version":"0.0.1"}
```
Comment (ericmj) [inline test_suite_analysis/metadata.json:1]: ✂️ ?

File: priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+defmodule Hexpm.RepoBase.Migrations.CreateStoredProcedureToPurgePackageSearches do
```
Comment (ericmj) [inline priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs:1]: We don't use stored procedures elsewhere in the repo. Any reason this cannot be an elixir function?

**Review by ericmj — APPROVED** [review]

File: priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs (hunk 3 above)
Comment (supersimple) [reply priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs:1]: That's cool. I think it just depends on how you want to trigger it.
Are there any other functions in the codebase that do something like this? (ie. where do they live?)
Would you want to set up a cron in GCP to run that function periodically?

Comment (ericmj) [reply priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs:1]: We can use https://github.com/citusdata/pg_cron, that extension is available on Cloud SQL or we can set up a cron in kubernetes that runs a release task https://github.com/hexpm/hexpm-ops/blob/main/kubernetes/prod/hexpm-stats.yml.

File: lib/hexpm/repository/package_searches/package_searches.ex [hunk 4]
```diff
@@ -0,0 +1,41 @@
+defmodule Hexpm.Repository.PackageSearches do
```
Comment (ericmj) [inline lib/hexpm/repository/package_searches/package_searches.ex:1]: Let's move this file up one directory to match the rest of the context files in this repo. https://elixiross.slack.com/archives/C02JXUJD4KG/p1717849774114129

File: priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs (hunk 3 above)
Comment (supersimple) [reply priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs:1]: cool. my first impression is to switch to a release task for visibility and testability. Adding a pg extension seems to be going down the more opaque path.
WDYT?

Comment (ericmj) [reply priv/repo/migrations/20240605015345_create_stored_procedure_to_purge_package_searches.exs:1]: Yep, sounds good

**Pull request merged**